addons/website_sale/controllers/main.py

```diff
@@ -242,24 +245,19 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
 
         Product = request.env['product.template'].with_context(bin_size=True)
 
-        Category = request.env['product.public.category']
-        search_categories = False
+        search_categories = Category
         search_product = Product.search(domain)
+        website_domain = request.website.website_domain()
         if search:
-            categories = search_product.mapped('public_categ_ids')
-            search_categories = Category.search([('id', 'parent_of', categories.ids)] + request.website.website_domain())
-            categs = search_categories.filtered(lambda c: not c.parent_id)
+            search_categories = Category.search([('product_tmpl_ids', 'in', search_product.ids)] + website_domain).parents_and_self
+            categs = Category.search(['|', ('id', 'in', search_categories.ids), ('id', 'child_of', search_categories.ids)] + website_domain)
```

Comment:
Just a last small remark as I was still on your branch while working on my PR for sitemap
I noticed that this seems not correct as it won't show categories of the product but the top parent only.
Just check screen shot, easier than writting :yum:

![Screenshot from 2019-06-24 17-02-16](https://user-images.githubusercontent.com/30048408/60029940-59fafc80-96a2-11e9-9fed-f5848df83ef8.png)
![Screenshot from 2019-06-24 17-02-14](https://user-images.githubusercontent.com/30048408/60029938-59fafc80-96a2-11e9-8908-b9e7014f3fd9.png)

![Screenshot from 2019-06-24 17-02-06](https://user-images.githubusercontent.com/30048408/60029936-59626600-96a2-11e9-9baf-baee64a188aa.png)
![Screenshot from 2019-06-24 17-02-04](https://user-images.githubusercontent.com/30048408/60029935-59626600-96a2-11e9-9b7c-bd3530785685.png)

Reply:
As discussed this was on purpose for performance (commit 2), but I'll remove it from this PR and maybe make another task, this could be improved for example in collapse mode only.